src/graph/graphcontext.c

```diff
@@ -148,47 +105,143 @@ LabelStore* GraphContext_AddRelationType(GraphContext *gc, const char *label) {
+}
 
+int GraphContext_AddIndex(GraphContext *gc, const char *label, const char *property) {
   // Find the ID of the specified label
   int label_id = GraphContext_GetLabelID(gc, label, STORE_NODE);
   if (label_id < 0 ) return INDEX_FAIL;
 
+  LabelStore *store = gc->node_stores[label_id];
+
+  // Triemaps (for some reason) don't specify const.
+  void *property_lookup = LabelStore_RetrieveValue(store, (char*)property);
+
+  if (property_lookup == TRIEMAP_NOTFOUND || property_lookup != NULL) {
+    // Property does not exist on this label or property is already indexed.
+    return INDEX_FAIL;
+  }
+
   // Populate an index for the label-property pair using the Graph interfaces.
   Index *idx = Index_Create(gc->g, label_id, label, property);
-  gc->indices[gc->index_count] = idx;
-  gc->index_count++;
+  gc->indices = array_append(gc->indices, idx);
+
+  // Associate the new index with the property in the label store.
+  LabelStore_AssignValue(store, (char*)property, (void*)idx);
 
   return INDEX_OK;
 }
 
 int GraphContext_DeleteIndex(GraphContext *gc, const char *label, const char *property) {
-  int offset = _GraphContext_IndexOffset(gc, label, property);
-  if (offset < 0) return INDEX_FAIL;
+  // Find the ID of the specified label
+  int label_id = GraphContext_GetLabelID(gc, label, STORE_NODE);
+  if (label_id < 0 ) return INDEX_FAIL;
 
-  Index *idx = gc->indices[offset];
+  LabelStore *store = gc->node_stores[label_id];
+
+  Index *idx = LabelStore_RetrieveValue(store, (char*)property);
+  if (idx == TRIEMAP_NOTFOUND || idx == NULL) {
+    // Property does not exist or was not indexed.
+    return INDEX_FAIL;
+  }
+
+  // Remove the index association from the label store
+  LabelStore_AssignValue(store, (char*)property, NULL);
+  // Index *idx = gc->indices[offset];
+
+  // Pop the last stored index
+  Index *last_idx = array_pop(gc->indices);
+  if (idx != last_idx) {
+    // If the index being deleted is not the last, swap the last into the newly-emptied position
+    uint32_t pos;
+    // Find the position of the index in the gc->indices array
+    for (pos = 0; pos < array_len(gc->indices); pos ++) {
+      if (gc->indices[pos] == idx) {
+        gc->indices[pos] = last_idx;
+        break;
+      }
+    }
+  }
+
+  // Free the index
   Index_Free(idx);
-  gc->index_count --;
-  // If the deleted index was not the last, swap the last into the newly-emptied position
-  if (offset < gc->index_count) gc->indices[offset] = gc->indices[gc->index_count];
 
   return INDEX_OK;
 }
 
+void GraphContext_AddNodeToIndices(GraphContext *gc, LabelStore *store, Node *n) {
+  if (store && GraphContext_HasIndices(gc)) {
+    EntityProperty *props = ENTITY_PROPS(n);
+    for (int j = 0; j < ENTITY_PROP_COUNT(n); j ++) {
+      Index *idx = LabelStore_RetrieveValue(store, props[j].name);
+      if (idx != TRIEMAP_NOTFOUND && idx != NULL) {
+        Index_InsertNode(idx, n->entity->id, &props[j].value);
+      }
+    }
+  }
+}
+
+// TODO inefficient
```

Comment:
Why is this inefficient?

Reply:
Hm, I had added this comment because it looks up the node store for every given property, rather than per node like the other functions. However, I guess that makes sense in the context of SET ops - only one property is getting update per expression anyway.